src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectAssetFileWatcher.cs

```diff
@@ -204,60 +220,88 @@ private void UnregisterFileWatcherIfAny()
             {
                 // There's nothing for us to do if this fails. So ignore the return value.
                 _fileChangeService?.UnadviseFileChange(_filechangeCookie);
+                _watchedFileResetCancellationToken?.Cancel();
+                _watchedFileResetCancellationToken?.Dispose();
+                _taskDelayScheduler?.Dispose();
             }
         }
 
-        protected override Task DisposeCoreAsync(bool initialized)
+        private CancellationToken CreateLinkedCancellationToken()
         {
-            if (initialized)
-            {
-                _treeWatcher.Dispose();
-                UnregisterFileWatcherIfAny();
-            }
+            // we want to cancel when we switch what file is watched, or when the project is unloaded
+            _watchedFileResetCancellationToken = CancellationTokenSource.CreateLinkedTokenSource(
+                _projectServices.Project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
 
-            return Task.CompletedTask;
+            return _watchedFileResetCancellationToken.Token;
         }
 
-        /// <summary>
-        /// Called when a project.lock.json file changes.
-        /// </summary>
-        public int FilesChanged(uint cChanges, string[] rgpszFile, uint[] rggrfChange)
+        private async Task HandleFileChangedAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
-            // Kick off the operation to notify the project change in a different thread irregardless of
-            // the kind of change since we are interested in all changes.
-            _projectServices.ThreadingService.Fork(async () =>
+            try
             {
-                try
+                string currentContents = null;
+                if (File.Exists(_fileBeingWatched))
                 {
+                    currentContents = File.ReadAllText(_fileBeingWatched);
```

Comment:
Need to handle more IO exceptions here.  `File.Exists` can throw, and even if it doesn't, this is a race condition.  The file could be deleted between these two calls.